Report rows_merged in compaction_history rest api and nodetool

paszkow opened the pull request:
Currently, running the `nodetool compactionhistory` command or using the rest api `curl -X GET --header "Accept: application/json" "http://localhost:10000/compaction_manager/compaction_history"` return compaction history without the `row_merged` field.

The series computes rows merged during compaction and provides this information to users via both the nodetool command and the rest api. The `rows_merged` field contains information on merged clustering keys across multiple sstable files. For instance, compacting two sstables of a table consisting of 7 rows where two rows are part of the both sstables, the output would have the following format: {1: 5, 2: 2}.

No backport is required. It extends the existing compaction history output.

Fixes https://github.com/scylladb/scylladb/issues/666

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11411/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11411/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: nodetool/test_compactionhistory
:small_blue_diamond: rest_api/test_compactionhistory
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11411/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3457/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/407/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11480/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11411/)
#### Build Details:
- Duration: 3 hr 5 min
- Builder: spider5.cloudius-systems.com

raphaelsc commented on mutation/mutation_compactor.hh:
```diff
@@ -683,6 +691,12 @@ public:
     }
 
     stop_iteration consume(clustering_row&& cr) {
+        if constexpr (SSTableCompaction == compact_for_sstables::yes) {
```
twcs uses compacting_reader, which bypasses this compact_mutation_v2 wrapper. so let's move it to compact_mutation_state.

raphaelsc commented on compaction/compaction.cc:
```diff
@@ -811,7 +812,8 @@ class compaction {
                         get_compacted_fragments_writer(),
-                        get_gc_compacted_fragments_writer());
+                        get_gc_compacted_fragments_writer(),
+                        &_rows_merged);
```
see that we missed make_compacting_reader() in consume_without_gc_writer(). I think we can add a test with TimeWindow usage.

raphaelsc commented on mutation/mutation_compactor.hh:
```diff
@@ -657,10 +661,14 @@ public:
+        , _clustering_keys_distribution(ck_distribution) {
+            if (_clustering_keys_distribution) {
+                _clustering_keys_distribution->clear();
+            }
```
I guess we don't have to clear the map here, but it doesn't hurt.

raphaelsc commented on mutation/mutation_fragment.hh:
```diff
@@ -112,6 +115,10 @@ public:
     }
 
+    int32_t merge_count() const {
+        return _merge_count;
```
nit: unsigned?

paszkow replied:
Yes, size_t would be better. Used int23_t though because there were already maps using this type in e.g. `system_keyspace::update_compaction_history`. This way I could focused first on covering all paths and leave this as the last step.

avikivity replied:
size_t is useful for counting in-memory objects. Rows are on-disk, so uint64_t is indicated.

avikivity replied:
Ah, that number corresponds to the number of sstables merged. That has to be a small number, it's not proportional to memory size.

avikivity commented on mutation/mutation_fragment.hh:
```diff
@@ -32,6 +32,7 @@
 class clustering_row {
     clustering_key_prefix _ck;
     deletable_row _row;
+    int32_t _merge_count{ 1 };
```
Why is the statistic in the row itself?

denesb replied:
That is simple enough, the tricky part is propagating up this statistics though all the reader layers, to the compaction one. The `clustering_row` is a natural place to store this IMO.

avikivity replied:
No, clustering_row has nothing to do with counting merges during compaction.

Propagating the statistics structure through the layers is annoying, but doesn't inject transient data where it doesn't belong. Don't be lazy.

avikivity commented on compaction/compaction.cc:
```diff
@@ -455,6 +455,7 @@ class compaction {
     uint64_t _estimated_partitions = 0;
     double _estimated_droppable_tombstone_ratio = 0;
     uint64_t _bloom_filter_checks = 0;
+    std::unordered_map<int32_t, int64_t> _rows_merged;
```
This is very opaque. Since it's a histogram, name it as such. Explain what the key is, since int32 doesn't say much.

It's also going to have a measurable impact on compaction speed. Better use one of our histogram types.

avikivity commented on test/rest_api/test_compactionhistory.py:
```diff
@@ -0,0 +1,53 @@
+
+
+def test_compactionhistory_rows_merged_task(cql, rest_api):
+    with new_test_keyspace(cql, "WITH replication = {'class': 'SimpleStrategy', 'replication_factor': 1}") as ks:
```
Don't use SimpleStrategy in new code, it's deprecated. Use NetworkTopologyStrategy.

paszkow commented:
In v2:

- Dropped commits introducing statistics to `clustering_row` and collecting histogram in `compact_mutation_v2` wrapper
- Propagate histogram down to merger_reader and populate it with data in `mutation_fragment_merger`
- Add additional tests to cover TWCS and a compaction of a single sstable file
- Renamed variables to let them have more meaningful names

denesb commented on readers/combined.cc:
```diff
@@ -83,12 +83,17 @@ class mutation_fragment_merger {
+    // An optional histogram describing a distribution of clustering keys.
+    // The map gathers a number of clustering keys merged (value) from a given
+    // number of sstable files (key).
+    std::unordered_map<int32_t, int64_t>* _histogram{ nullptr };
```
We have histogram types in `util/histogram.hh`, please see if one of those are a good fit here.

avikivity replied:
Absolutely

avikivity replied:
I think it's a good idea to wrap it in a `struct combined_reader_statistics`, to clarify the intent, and make room for more options. It's probably faster to make it non-optional.

paszkow replied:
Agree. I realised there is a second associated task that requires gathering more information from the readers. Will wrap in an additional struct. 
I made it optional, as this information is only required by a regular compaction. If it is not a problem to collect data with every call to the make_combined_reader, then it can be non-optional as well.

paszkow replied:
> We have histogram types in `util/histogram.hh`, please see if one of those are a good fit here.

Looked at the file and did not find any suitable types for this purpose. Thus I left the map. I think a simple map is a better solution for this purpose. Especially when wrapped in an additional structure.

avikivity replied:
On modern processors additions are free but if ()s are expensive.

The map is a terrible structure for this. Compared to the addition, the lookup and possible allocation are very expensive.

A good type for this is `approx_estimated_histogram<8, 64, 8>`. It will lookup the bucket using just a few instructions.

paszkow replied:
According to the cassandra's nodetool compactionhistory [documentation](https://docs.datastax.com/en/cassandra-oss/3.x/cassandra/tools/toolsCompactionHistory.html), the statistic in the `rows_merged` field has a notation `{tables:rows}`, e.g. `{1:3, 3:1}` means 3 rows were taken from one SSTable (1:3) and 1 row taken from 3 SSTables (3:1). 

If we want to stick to the same format, then a hash map I used in the implementation is sufficient for that. 

On the other hand, if we want a proper histogram and aggregate data into buckets then I agree `approx_exponential_histogram` could be a better choice. That will result in a different output, however. Take the example above. The values 1 and 3 can get into one bucket and then its value would be 4 and we would have something like `{[1:4): 4}`

avikivity replied:
We can make it simple. We know in advance how many sstables we are compacting, so we can allocate a vector of size n+1 and increment the appropriate slot.

denesb commented on sstables/sstable_set.cc:
```diff
@@ -1380,7 +1381,7 @@ sstable_set::make_local_shard_sstable_reader(
         return reader;
     };
-    if (_impl->size() == 1) [[unlikely]] {
+    if (_impl->size() == 1 && !histogram) [[unlikely]] {
```
I don't think `[[unlikely]]` is justified here. This function is used by user reads too, which will never pass `histogram` here.

denesb commented on compaction/compaction.hh:
```diff
@@ -77,6 +77,7 @@ struct compaction_stats {
     uint64_t validation_errors = 0;
     // Bloom filter checks during max purgeable calculation
     uint64_t bloom_filter_checks = 0;
+    std::unordered_map<int32_t, int64_t> histogram;
```
Needs a more specific name, than just `histogram`.

denesb reviewed (changes requested):
We need a more specific name for the variable passed around, than just `histogram`. `rows_merged_histogram` would be better.

avikivity commented on compaction/compaction.cc:
```diff
@@ -1105,7 +1110,9 @@ class regular_compaction : public compaction {
-                unwrap_monitor_generator());
+                unwrap_monitor_generator(),
+                default_sstable_predicate(),
+                &_histogram);
```
@raphaelsc why is this const?

raphaelsc replied:
probably it made sense initially, but now I see that even scrub_compaction mutates its internal member in the reader returned.

avikivity commented on test/nodetool/test_compactionhistory.py:
```diff
@@ -36,27 +37,28 @@ def format_compacted_at(compacted_at: int):
 id                                   keyspace_name columnfamily_name compacted_at            bytes_in bytes_out rows_merged
-edef82f0-5e9c-11ee-a8f6-7d85dcfeb8f4 system_schema functions         {}     5790      5944            
-edde9300-5e9c-11ee-a8f6-7d85dcfeb8f4 system        peers             {}    11714     11808            
+edef82f0-5e9c-11ee-a8f6-7d85dcfeb8f4 system_schema functions         {} 5790     5944  {{2: 1, 1: 5}}
```
Let's sort the output by the key.

avikivity commented on test/nodetool/test_compactionhistory.py:
```diff
@@ -80,7 +82,16 @@ def test_json(nodetool):
+                        {
+                            "key": 1,
+                            "value": 5
+                        }
```
Why expose a map as an array of key-value objects, and not as a map?

avikivity replied:
Ah, maybe the schema is preexisting.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11591/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11591/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: nodetool/test_compactionhistory
:small_blue_diamond: rest_api/test_compactionhistory
#### Failed Tests (22/270):
- [test_compactionhistory_rows_merged_null_compaction_strategy_one_sstable_file.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11591/testReport/junit/%28root%29/test_compactionhistory/test_compactionhistory_rows_merged_null_compaction_strategy_one_sstable_file_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compactionhistory_rows_merged_null_compaction_strategy_one_sstable_file.debug.1)
- [test_compactionhistory_rows_merged_null_compaction_strategy_one_sstable_file.debug.10](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11591/testReport/junit/%28root%29/test_compactionhistory/test_compactionhistory_rows_merged_null_compaction_strategy_one_sstable_file_debug_10) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compactionhistory_rows_merged_null_compaction_strategy_one_sstable_file.debug.10)
- [test_compactionhistory_rows_merged_null_compaction_strategy_one_sstable_file.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11591/testReport/junit/%28root%29/test_compactionhistory/test_compactionhistory_rows_merged_null_compaction_strategy_one_sstable_file_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compactionhistory_rows_merged_null_compaction_strategy_one_sstable_file.debug.2)
- [test_compactionhistory_rows_merged_null_compaction_strategy_one_sstable_file.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11591/testReport/junit/%28root%29/test_compactionhistory/test_compactionhistory_rows_merged_null_compaction_strategy_one_sstable_file_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compactionhistory_rows_merged_null_compaction_strategy_one_sstable_file.debug.3)
- [test_compactionhistory_rows_merged_null_compaction_strategy_one_sstable_file.debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11591/testReport/junit/%28root%29/test_compactionhistory/test_compactionhistory_rows_merged_null_compaction_strategy_one_sstable_file_debug_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compactionhistory_rows_merged_null_compaction_strategy_one_sstable_file.debug.4)
- [test_compactionhistory_rows_merged_null_compaction_strategy_one_sstable_file.debug.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11591/testReport/junit/%28root%29/test_compactionhistory/test_compactionhistory_rows_merged_null_compaction_strategy_one_sstable_file_debug_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compactionhistory_rows_merged_null_compaction_strategy_one_sstable_file.debug.5)
- [test_compactionhistory_rows_merged_null_compaction_strategy_one_sstable_file.debug.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11591/testReport/junit/%28root%29/test_compactionhistory/test_compactionhistory_rows_merged_null_compaction_strategy_one_sstable_file_debug_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compactionhistory_rows_merged_null_compaction_strategy_one_sstable_file.debug.6)
- [test_compactionhistory_rows_merged_null_compaction_strategy_one_sstable_file.debug.7](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11591/testReport/junit/%28root%29/test_compactionhistory/test_compactionhistory_rows_merged_null_compaction_strategy_one_sstable_file_debug_7) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compactionhistory_rows_merged_null_compaction_strategy_one_sstable_file.debug.7)
- [test_compactionhistory_rows_merged_null_compaction_strategy_one_sstable_file.debug.8](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11591/testReport/junit/%28root%29/test_compactionhistory/test_compactionhistory_rows_merged_null_compaction_strategy_one_sstable_file_debug_8) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compactionhistory_rows_merged_null_compaction_strategy_one_sstable_file.debug.8)
- [test_compactionhistory_rows_merged_null_compaction_strategy_one_sstable_file.debug.9](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11591/testReport/junit/%28root%29/test_compactionhistory/test_compactionhistory_rows_merged_null_compaction_strategy_one_sstable_file_debug_9) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compactionhistory_rows_merged_null_compaction_strategy_one_sstable_file.debug.9)
#### Build Details:
- Duration: 43 min
- Builder: spider4.cloudius-systems.com

paszkow commented:
In v4:

- Rebased
- Fix a failing test in the Debug mode

raphaelsc commented on readers/combined.cc:
```diff
@@ -689,42 +691,47 @@ mutation_reader make_combined_reader(schema_ptr schema,
     }
-    if (readers.size() == 1) {
+    if (readers.size() == 1 && !statistics) {
         return std::move(readers.front());
```
should we rule out this optimization based on the presence of stats? I think we shouldn't.

paszkow replied:
The statistics are collected in the combined reader. I use `mutation_fragment_batch` to get the number of sstables the row is read from. Thus I skip the optimization. Otherwise, I would need to propagate the statistics structure to a `mutation_reader` and I do not think it is necessary.

Note that the optimization is skipped only for compaction when statistics are required. For the rest of the reads, the optimization is not omitted.

raphaelsc commented on sstables/sstable_set.cc:
```diff
@@ -1381,7 +1382,7 @@ sstable_set::make_local_shard_sstable_reader(
     };
-    if (_impl->size() == 1) [[unlikely]] {
+    if (_impl->size() == 1 && !statistics) {
         auto sstables = _impl->all();
```
ditto.

raphaelsc replied:
also, with a single reader, all rows will come from a single sstable. I suggest we leave the stats unfilled like today if compaction works with a single sstable.

raphaelsc replied:
so one can interpret it as no merging across sstables happened.

paszkow replied:
> we cannot afford to lose the optimization, the stats is less important, see: [f75154a](https://github.com/scylladb/scylladb/commit/f75154afca472251391445697202d980a510c9a5)

Right. Definitely sounds reasonable to keep the optimization. Then, as you proposed, I will leave the field empty.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12133/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12133/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: nodetool/test_compactionhistory
:small_blue_diamond: rest_api/test_compactionhistory
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12133/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3993/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/938/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12015/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12133/)
#### Build Details:
- Duration: 3 hr 48 min
- Builder: spider5.cloudius-systems.com

bhalevy commented:
@paszkow is this PR ready for merging?
If so, please add the `status/marge candidate` label.

denesb commented:
> @paszkow is this PR ready for merging? If so, please add the `status/marge candidate` label.

There is still a discussion going on with @avikivity, see https://github.com/scylladb/scylladb/pull/20481?notification_referrer_id=NT_kwDOABUy2bMxMjI4OTI4MjY1MzoxMzg5Mjcz&notifications_query=reason%3Aparticipating#discussion_r1794074718.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12333/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12333/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: nodetool/test_compactionhistory
:small_blue_diamond: rest_api/test_compactionhistory
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12333/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1087/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4150/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12173/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12333/)
#### Build Details:
- Duration: 6 hr 23 min
- Builder: i-0daa1be0e8fa96b10 (m5ad.12xlarge)

paszkow commented:
In v6:

- Switched from using a hashmap into a simple vector to collect rows merged statistics. That required more changes in the first commint ([317baed](https://github.com/scylladb/scylladb/pull/20481/commits/317baed95558d448cfda69100525c08928f7cfb6)) in the series.

raphaelsc commented on readers/combined.cc:
```diff
@@ -114,6 +116,10 @@ class mutation_fragment_merger {
                 } else {
+                    if (_statistics && begin->fragment.is_clustering_row()) {
+                        ++_statistics->rows_merged_histogram[fragments.size()];
+                    }
```
maybe we should treat the caller resizing the vector upfront as optimization and optionally grow the histogram here (slower but still works). today it works, but if we have another user tomorrow that forgets to resize the vector (or the resizing in the current user goes away by mistake), we won't have surprises. Usually it's better to not have such contracts between caller and callees. if we go that way, I suggest hiding it in a function to avoid polluting this main function.

denesb replied:
Agree, we should make sure the code is safe, the pre-set size is an optimization, let's not rely on it.

paszkow replied:
The counter is increased in `mutation_fragment_merger` which does not have any knowledge on the number of sstable files involved. Thus I cannot properly resize the vector there even though I would really like.

The `compation` class (the caller) that requests the statistics to be collected has this knowledge and should ensure the vector is of a proper size.

If we want to check for the out-of-bound index and optionally resize the vector, then I insist to take a step back and use a hashmap. It does it for us and we can still do some optimizations by calling `reserve` on it to avoid unnecessary allocations.

avikivity replied:
I think it's okay to require the caller and callee to cooperate here. This is going to be called hundreds of thousands of times per second in some cases, we can't start hashing things in hot paths like that for no reason.

avikivity commented:
> 
> Fixes #666

Auspicious

avikivity commented on db/system_keyspace.cc:
```diff
@@ -2352,19 +2352,20 @@ system_keyspace::query(distributed<replica::database>& db, const sstring& ks_nam
-        auto map_element = std::make_pair<data_value, data_value>(data_value(first), data_value(second));
+    for (size_t id=0; id<rows_merged.size(); ++id) {
+        if (rows_merged[id] <= 0)
+            continue;
```
Our coding style requires braces for single-line statements in control flow.

avikivity commented on readers/combined.cc:
```diff
@@ -114,6 +116,10 @@ class mutation_fragment_merger {
                     }
                     return stop_iteration::no;
                 } else {
+                    if (_statistics && begin->fragment.is_clustering_row()) {
```
I'd drop the check for is_clustering_row() and move the accounting before the `if`. After that change, we don't account for just clustering rows, but also static rows, partition starts, range tombstones, etc. It dilutes the meaning of "rows" in rows merged, but it's a more useful metric since it tells you how much data was merges vs not merged, regardless of exact type, which isn't very interesting.

avikivity commented on readers/combined.hh:
```diff
@@ -49,15 +49,18 @@ mutation_reader make_combined_reader(schema_ptr schema,
         streamed_mutation::forwarding fwd_sm = streamed_mutation::forwarding::no,
-        mutation_reader::forwarding fwd_mr = mutation_reader::forwarding::yes);
+        mutation_reader::forwarding fwd_mr = mutation_reader::forwarding::yes,
+        combined_reader_statistics* statistics = nullptr);
```
Wouldn't this be the right spot to allocate the vector? Or at least resize it? To avoid a mismatch.

paszkow replied:
I can definitely resize it here but that is not enough as `sstable_set::make_local_shard_sstable_reader` calls the variant with `std::unique_ptr<reader_selector>`.

So instead, the plan is to extend `reader_selector` with the information on the maximum readers count and use it to resize the statistics vector. This way, it will be done exactly in one place that actually creates the reader.

avikivity replied:
Ok. Note that having multiple resizes isn't a problem, as long as it's not in the fast path.

avikivity commented on compaction/compaction.cc:
```diff
@@ -826,6 +827,7 @@ class compaction {
         _estimated_droppable_tombstone_ratio = std::min(1.0, sum_of_estimated_droppable_tombstone_ratio / ssts->size());
 
         _compacting = std::move(ssts);
+        _reader_statistics.rows_merged_histogram.resize(_compacting->size()+1, 0);
```
As noted before, we should let make_combined_reader (or perhaps its immediate callee) resize the vector, since it is the one who is intimately familiar with how it will be used. To be safe, it should only resize it upwards.

avikivity commented on compaction/compaction.hh:
```diff
@@ -9,6 +9,7 @@
 
 #pragma once
 
+#include "readers/combined.hh"
```
This is adding a dependency of the whole reader hierarchy. Let's split of the stats struct into its own header file to reduce rebuilds.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:x: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12390/)
#### Build Details:
- Duration: 1 hr 28 min
- Builder: i-0444a5f7eb3083afb (m5ad.8xlarge)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12398/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12398/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/mutation_reader_test
:small_blue_diamond: nodetool/test_compactionhistory
:small_blue_diamond: rest_api/test_compactionhistory
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12398/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4193/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12216/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1131/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12398/)
#### Build Details:
- Duration: 4 hr 23 min
- Builder: spider2.cloudius-systems.com

paszkow commented:
In v8:

- Rebased
- Dropped the commit 8bc9b966a4dd7ec6. The `rows_merged` column of the `compaction_history` table, the `compaction_history_entry` struct and the API, they all expect a map. Furthermore, the next PR that is about to be sent extending the `compaction_history` table, changes `system_keyspace::update_compaction_history` to re-use `compaction_history_entry` struct. So instead of reverting the commit later, it is better to drop it now.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12647/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12647/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/mutation_reader_test
:small_blue_diamond: nodetool/test_compactionhistory
:small_blue_diamond: rest_api/test_compactionhistory
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12647/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1302/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12390/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4369/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12647/)
#### Build Details:
- Duration: 9 hr 1 min
- Builder: i-0b44df747e5ae2fa6 (m5ad.8xlarge)
